common/src/refcount.rs

```diff
@@ -0,0 +1,52 @@
+use crate::atomic::{Ordering::*, PyAtomic, Radium};
+
+/// from alloc::sync
+/// A soft limit on the amount of references that may be made to an `Arc`.
+///
+/// Going above this limit will abort your program (although not
+/// necessarily) at _exactly_ `MAX_REFCOUNT + 1` references.
+const MAX_REFCOUNT: usize = (isize::MAX) as usize;
+
+pub struct RefCount {
+    strong: PyAtomic<usize>,
+}
+
+impl Default for RefCount {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RefCount {
+    pub fn new() -> Self {
+        RefCount {
+            strong: Radium::new(1),
+        }
+    }
+
+    #[inline]
+    pub fn get(&self) -> usize {
+        self.strong.load(SeqCst)
+    }
+
+    #[inline]
+    pub fn incref(&self) {
```

Comment:
Idk, I think I'd like us to use "incref" more in the codebase - like for the other borrowed PyObject change, I used incref because in general I think it's weird to `clone()`/`to_owned()` stuff that isn't actually owned separately. I can think of one instance where someone commenting on a PR thought that an issue was caused by `clone()` creating a new object, even though it doesn't. (This I think is why `Rc`/`Arc` encourages `Rc::clone(&foo)` instead of `foo.clone()` because it makes it always clear that it's actually just incr'ing a refcount.) So I'd like to switch to `incref()` instead of to_owned/clone at some point (another reason it was specifically in that pr - it makes so you don't have to think about whether the object is borrowed and you should do to_owned or owned (clone), if you want to incref you always just do incref()) and I think it makes sense to implement incref by calling an inner incref. (long explanation lol.) another thing ig is that cpython uses "incref" a lot, even the lowercase word (not the Py_INCREF macro) is used 50+ times just in comments/prose.

Reply:
The suggestion was not about incref/clone but I am not sure that's real. Do you have a reference that Rc/Arc encourages `Rc::clone`? As I know, they generally use it to distinguish `Rc<T>::clone` and `T::clone` because both of them are accessible via `Deref`, not because they give any cost sense.

I don't think "cpython uses the name" can be a good support for a name for now. They are choosing a lot of bad namings in Rust convention point of view and we already are not following it that much. At least, we mostly are adding `_`s unless it is a Python symbol.

Reply:
> Do you have a reference that Rc/Arc encourages `Rc::clone`?

https://doc.rust-lang.org/stable/std/rc/index.html

> `Rc<T>`’s implementations of traits like Clone may also be called using fully qualified syntax. Some people prefer to use fully qualified syntax, while others prefer using method-call syntax.

> The `Rc::clone(&from)` syntax is the most idiomatic because it conveys more explicitly the meaning of the code. In the example above, this syntax makes it easier to see that this code is creating a new reference rather than copying the whole content of foo.

Reply:
Oh, thanks. I didn't know that. The example looks like to choose `Rc<Vec>` to show the difference between `Vec::clone` and `Rc::clone`. Still I am not sure how it supports `incref` instead of `clone`.